src/simulators/qasm/qasm_controller.hpp

```diff
@@ -299,8 +322,12 @@ void QasmController::initialize_state(const Circuit &circ,
 template <class State_t>
 void QasmController::optimize_circuit(const Circuit &input_circ,
```

Comment:
Maybe we should also change the function definition to be:

```c++
Circuit optimize_circuit(const Circuit &circ) const
```

So that it returns the output circuit object and is used like

```c++
auto opt_circ = optimize_circuit<State_t>(circ);
```

rather than 

```c++
Circuit opt_circ;
optimize_circuit<State_t>(circ, opt_circ);
```